app/assets/javascripts/master_detail.js

```diff
@@ -89,48 +89,62 @@ function clickBack(btn, level) {
 
   try {
-    if($grid.data("compact-view") && ($grid.data("compact-view")[level] || []).length > 0) {
-      $.each($grid.data("compact-view")[level], function(i, item) {
+    preserve_selected_rows = $grid.data("master_detail-preserve_selected_rows");
+    preserve_selected_rows = preserve_selected_rows === undefined ? false : preserve_selected_rows;
+    apply_default_state_on_reset = $grid.data("master_detail-apply_default_state_on_reset");
+    apply_default_state_on_reset = apply_default_state_on_reset === undefined ? true : apply_default_state_on_reset;
+    perform_update_dimensions_and_repaint_in_timeout = $grid.data("master_detail-perform_update_dimensions_and_repaint_in_timeout");
+    perform_update_dimensions_and_repaint_in_timeout = perform_update_dimensions_and_repaint_in_timeout === undefined ? true : perform_update_dimensions_and_repaint_in_timeout;
+
+    if (preserve_selected_rows) {
+      selectedRowKeys = $grid.dxDataGrid('instance').getSelectedRowKeys();
+    }
+
+    if ($grid.data("compact-view") && ($grid.data("compact-view")[level] || []).length > 0) {
+      $.each($grid.data("compact-view")[level], function (i, item) {
         $grid.dxDataGrid('columnOption', item.name, item.property, item.value);
       });
+    } else if (factory_reset && apply_default_state_on_reset) {
+      if ($grid.data("default-state-json") !== undefined) {
+        $grid.dxDataGrid('instance').state($grid.data("default-state-json"));
+      } else {
+        $grid.dxDataGrid({
+          columns: $grid.data("default-json")
+        });
+      }
     }
   }
   catch(e) {
     // Catches the instance where the compact view is not defined in a master detail (but its indeterminable)
   }
   finally {
-    window.setTimeout(function(){
-      // If someone is clicking fast on different rows, the grid may not be on the dom anymore.
-      // We don't want to repaint the grid if it isn't there. Plus, it will fail if it isn't there.
-      if ($(selector).length > 0) {
-        dataGrid = $grid.dxDataGrid('instance');
-        dataGrid.updateDimensions();
-        dataGrid.repaint();
-      }
-    }, 1000);
-  }
-}
-
-function factory_reset_grid($grid){
-  if ($grid.data("default-state-json") !== undefined) {
-    $grid.dxDataGrid('instance').state($grid.data("default-state-json"));
-  } else {
-    $grid.dxDataGrid({
-      columns: $grid.data("default-json")
-    });
+    if (perform_update_dimensions_and_repaint_in_timeout) {
+      window.setTimeout(function () {
+        // If someone is clicking fast on different rows, the grid may not be on the dom anymore.
+        // We don't want to repaint the grid if it isn't there. Plus, it will fail if it isn't there.
+        if ($(selector).length > 0) {
+          dataGrid = $grid.dxDataGrid('instance');
+          dataGrid.updateDimensions();
+          dataGrid.repaint();
+          if (preserve_selected_rows) {
+            dataGrid.selectRows(selectedRowKeys, false);
```

Comment:
Consider setting `selectedRowKeys` directly on the datagrid instead of using this async method, if you want the changes to appear immediately.